libfwupdplugin/fu-bios-attrs.c

```diff
@@ -0,0 +1,424 @@
+ * @self: a #FuBiosAttrs
+ *
+ * Clears all attributes and re-initializes them.
+ * Mostly used for the test suite, but could potentially be connected to udev
+ * events for drivers being loaded or unloaded too.
+ *
+ * Since: 1.8.4
+ **/
+gboolean
+fu_bios_attrs_setup(FuBiosAttrs *self, GError **error)
+{
+	g_autofree gchar *sysfsfwdir = NULL;
+	g_autofree gchar *tmp = NULL;
+	g_autoptr(GDir) class_dir = NULL;
+
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), FALSE);
+
+	if (self->attrs->len > 0) {
+		g_debug("re-initializing attributes");
+		g_ptr_array_set_size(self->attrs, 0);
+	}
+
+	sysfsfwdir = fu_path_from_kind(FU_PATH_KIND_SYSFSDIR_FW_ATTRIB);
+	class_dir = g_dir_open(sysfsfwdir, 0, error);
+	if (class_dir == NULL)
+		return FALSE;
+
+	do {
+		g_autofree gchar *path = NULL;
+		g_autoptr(GDir) driver_dir = NULL;
+		const gchar *driver = g_dir_read_name(class_dir);
+		if (driver == NULL)
+			break;
+		path = g_build_filename(sysfsfwdir, driver, "attributes", NULL);
+		driver_dir = g_dir_open(path, 0, error);
+		if (driver_dir == NULL)
+			return FALSE;
+		do {
+			const gchar *name = g_dir_read_name(driver_dir);
+			g_autoptr(GError) error_local = NULL;
+			if (name == NULL)
+				break;
+			if (!fu_bios_attrs_populate_attribute(self,
+							      driver,
+							      path,
+							      name,
+							      &error_local)) {
+				if (g_error_matches(error_local,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_NOT_SUPPORTED)) {
+					g_debug("%s is not supported", name);
+					continue;
+				}
+				g_propagate_error(error, g_steal_pointer(&error_local));
+				return FALSE;
+			}
+		} while (TRUE);
+	} while (TRUE);
+
+	return TRUE;
+}
+
+static void
+fu_bios_attrs_init(FuBiosAttrs *self)
+{
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+	self->attrs = g_ptr_array_new_with_free_func((GDestroyNotify)g_object_unref);
+}
+
+/**
+ * fu_bios_attributes_get_attr:
+ * @self: a #FuBiosAttrs
+ * @name: the attribute "name" to check for
+ *
+ * Returns: (transfer none): the attribute with the given name or NULL if it doesn't exist.
+ *
+ * Since: 1.8.4
+ **/
+FwupdBiosAttr *
+fu_bios_attributes_get_attr(FuBiosAttrs *self, const gchar *name)
+{
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), NULL);
+	g_return_val_if_fail(name != NULL, NULL);
+
+	for (guint i = 0; i < self->attrs->len; i++) {
+		FwupdBiosAttr *attr = g_ptr_array_index(self->attrs, i);
+		const gchar *tmp = fwupd_bios_attr_get_name(attr);
+		if (g_strcmp0(tmp, name) == 0)
+			return attr;
+	}
+	return NULL;
+}
+
+/**
+ * fu_bios_attrs_get_all:
+ * @self: a #FuBiosAttrs
+ *
+ * Gets all the attributes in the object.
+ *
+ * Returns: (transfer container) (element-type FwupdBiosAttr): attributes
+ *
+ * Since: 1.8.4
+ **/
+GPtrArray *
+fu_bios_attrs_get_all(FuBiosAttrs *self)
+{
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), NULL);
+	return g_ptr_array_ref(self->attrs);
+}
+
+/**
+ * fu_bios_attributes_get_pending:
+ * @self: a #FuBiosAttrs
+ * @result: (out): Whether a reboot is pending
```

Comment:
could this be (nullable) too?

Reply:
No - 
	g_return_val_if_fail(result != NULL, FALSE);